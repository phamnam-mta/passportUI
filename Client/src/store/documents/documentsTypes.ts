export enum DocumentStatus {
    Loading = "Loading",
    Loaded = "Loaded",
    Analyzing = "Analyzing",
    Analyzed = "Analyzed",
    AnalyzeFailed = "AnalyzeFailed",
    Labeled = "Labeled",
}

export interface IDocumentStates {
    loadingStatus: DocumentStatus.Loaded | DocumentStatus.Loading;
    analyzingStatus?: DocumentStatus.Analyzed | DocumentStatus.Analyzing | DocumentStatus.AnalyzeFailed;
    labelingStatus?: DocumentStatus.Labeled;
}

export interface IRawDocument {
    name: string;
    type: string;
    url: string;
}

export interface IDocument extends IRawDocument {
    thumbnail: string;
    numPages: number;
    currentPage: number;
    states: IDocumentStates;
    expirationTime?: number;
}
