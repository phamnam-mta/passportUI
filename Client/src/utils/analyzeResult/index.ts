import { BoundingRegion, Polygon, StudioFieldValue, StudioSpan } from "models/analyzeResult";

export const getPagePolygons = (boundingRegions: BoundingRegion[], page: number): Polygon[] => {
    return boundingRegions
        .filter((boundingRegion) => boundingRegion.pageNumber === page)
        .map((boundingRegion) => boundingRegion.polygon);
};

export const isAnyBoundingRegionInPage = (boundingRegions: BoundingRegion[], page: number): boolean => {
    return boundingRegions.some((boundingRegion) => boundingRegion.pageNumber === page);
};

export const getFieldTypeValue = (field: StudioFieldValue) => {
    const { type } = field;
    const typeString = type.charAt(0).toUpperCase() + type.slice(1);

    // TODO: Update when going public.
    if (type === "currency" && field.valueCurrency && typeof field.valueCurrency !== "string") {
        const { currencySymbol, currency, amount } = field.valueCurrency;
        if (currencySymbol && amount) {
            return `${currencySymbol}${amount}`;
        } else if (currency && amount) {
            return `${currency}${amount}`;
        } else if (amount) {
            return `${amount}`;
        } else {
            return undefined;
        }
    } else if (type === "address" && field.valueAddress) {
        return field.content;
    }
    return field[`value${typeString}`];
};

export interface SpanPosition extends StudioSpan {
    iconName: string;
    pageNumber?: number;
    tableIndex?: number;
    shouldDisplayAsIcon: boolean;
}
