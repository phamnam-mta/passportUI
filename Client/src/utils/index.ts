export const delay = (ms: number) => {
    return new Promise<void>((resolve) => {
        setTimeout(() => {
            resolve();
        }, ms);
    });
};

export const getPixelWidthFromPercent = (widthInPercentage: number): number => {
    const totalWindowWidth = document.documentElement.clientWidth;
    const paneWidthInPixel = (totalWindowWidth / 100) * widthInPercentage;
    return paneWidthInPixel;
};

export const loadCanvasToBlob = (canvas: HTMLCanvasElement): Promise<Blob> => {
    return new Promise((resolve, reject) => {
        canvas.toBlob((blob) => {
            if (blob) {
                resolve(blob);
            } else {
                reject("Failed to create blob for canvas.");
            }
        });
    });
};

export const loadUrlToArrayBuffer = async (url: string): Promise<ArrayBuffer> => {
    return await fetch(url).then((r) => r.arrayBuffer());
};

export const debounce = (func, timeout = 300) => {
    return (...args) => {
        return setTimeout(() => {
            func.apply(this, args);
        }, timeout);
    };
};
