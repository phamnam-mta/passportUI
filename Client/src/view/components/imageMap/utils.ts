import { Feature } from "ol";
import { Extent } from "ol/extent";
import Polygon from "ol/geom/Polygon";
import ImageLayer from "ol/layer/Image";
import Layer from "ol/layer/Layer";
import VectorLayer from "ol/layer/Vector";
import Projection from "ol/proj/Projection";
import ImageSource from "ol/source/ImageStatic";
import VectorSource from "ol/source/Vector";
import { LAYER_NAME } from "./constants";
import { IRegion } from "./contracts";

export type Dimension = {
    width: number;
    height: number;
};

export const degreeToRadians = (degree: number) => {
    // convert degree to radians
    return (degree * Math.PI * 2) / 360;
};

export const makeLayerFilter = (layerName) => {
    return {
        layerFilter: (layer: Layer) => layer.get(LAYER_NAME) === layerName,
    };
};

export const makeVectorLayer = (layerName, options = {}): Layer => {
    return new VectorLayer({
        [LAYER_NAME]: layerName,
        source: new VectorSource(),
        ...options,
    } as any);
};

export const makeImageLayer = (layerName, uri, projection, imageExtent, options = {}): Layer => {
    return new ImageLayer({
        [LAYER_NAME]: layerName,
        source: makeImageSource(uri, projection, imageExtent),
        ...options,
    } as any);
};

export const makeImageSource = (url: string, projection: Projection, imageExtent: Extent) => {
    return new ImageSource({ url, projection, imageExtent });
};

export const convertRegionToFeature = (region: IRegion, imageExtent: Extent, isOcrProposal: boolean = false) => {
    const coordinates: Array<[number, number]> = [];
    const boundingBox = region.id.split(",").map(parseFloat);
    const imageWidth = imageExtent[2] - imageExtent[0];
    const imageHeight = imageExtent[3] - imageExtent[1];
    for (let i = 0; i < boundingBox.length; i += 2) {
        coordinates.push([Math.round(boundingBox[i] * imageWidth), Math.round((1 - boundingBox[i + 1]) * imageHeight)]);
    }
    const feature = new Feature({
        geometry: new Polygon([coordinates]),
    });
    feature.setProperties({
        id: region.id,
        text: region.value,
        highlighted: false,
        isOcrProposal,
    });
    feature.setId(region.id);

    return feature;
};

export const createRegionIdFromPolygon = (polygon: number[], page: number): string => {
    return polygon.join(",") + ":" + page;
};

export const convertToImageMapCoordinates = (polygon: number[], canvasSize: Dimension, documentSize: Dimension) => {
    const { width: canvasWidth, height: canvasHeight } = canvasSize;
    const { width: ocrWidth, height: ocrHeight } = documentSize;

    const coordinates: number[][] = [];
    for (let i = 0; i < polygon.length; i += 2) {
        coordinates.push([(polygon[i] / ocrWidth) * canvasWidth, (1 - polygon[i + 1] / ocrHeight) * canvasHeight]);
    }
    return coordinates;
};

export const getBoundingBoxFromFeatureId = (id) => id.split(",").map(parseFloat);
