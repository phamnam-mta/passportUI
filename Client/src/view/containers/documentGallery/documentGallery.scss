@import "~@fluentui/react/dist/sass/References";

.document-gallery-container {
    display: flex;
    flex-direction: column;
    flex-grow: 1;

    .document-gallery-toolbar {
        margin: 8px 32px 8px 16px;
        display: flex;
        flex: 0 48px;
    }

    .document-gallery-list {
        display: flex;
        flex-grow: 1;
        height: 0;
    }
}
