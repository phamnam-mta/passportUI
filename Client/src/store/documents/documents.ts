import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { resetPredictions } from "store/predictions/predictions";
import { DocumentLoaderFactory, IDocumentLoader } from "utils/documentLoader";
import { DocumentStatus, IDocument, IRawDocument } from "./documentsTypes";

export type DocumentsState = { documents: IDocument[]; currentDocument: IDocument | null };

export const initialState: DocumentsState = { documents: [], currentDocument: null };

export const addDocuments = createAsyncThunk("documents/addDocuments", async (docsToAdd: IRawDocument[]) => {
    return Promise.all(docsToAdd.map(async (doc) => (await getLoader(doc)).loadDocumentMeta()));
});

export const setCurrentDocument = createAsyncThunk("documents/setCurrentDocument", async (document: IDocument) => {
    const loader = await getLoader(document);
    const documentPage = await loader.loadDocumentPage(document.currentPage);
    return { document, documentPage };
});

export const setCurrentPage = createAsyncThunk("documents/setCurrentPage", async (pageNumber: number, thunkApi) => {
    const { documents: state } = thunkApi.getState() as any;
    const loader = await getLoader(state.currentDocument);
    const documentPage = await loader.loadDocumentPage(pageNumber);
    return { pageNumber, documentPage };
});

const documentLoaders = new Map<string, IDocumentLoader>();

const getLoader = async (document: IRawDocument): Promise<IDocumentLoader> => {
    let loader = documentLoaders.get(document.url);
    if (!loader) {
        loader = await DocumentLoaderFactory.makeLoader(document);
        documentLoaders.set(document.url, loader);
    }

    return loader;
};

const documentsSlice = createSlice({
    name: "documents",
    initialState,
    reducers: {
        deleteDocument(state, action) {
            const docNameToDelete = action.payload.name;
            state.documents = state.documents.filter((document) => document.name !== docNameToDelete);
            documentLoaders.delete(action.payload.url);
        },
        setDocumentAnalyzingStatus(state, action) {
            const { name, status } = action.payload;
            const document = state.documents.find((document) => document.name === name);
            if (document) {
                document.states.analyzingStatus = status;
            }
            if (state.currentDocument?.name === name) {
                state.currentDocument!.states.analyzingStatus = status;
            }
        },
        setDocumentLabelingStatus(state, action) {
            const { name, status } = action.payload;
            const document = state.documents.find((document) => document.name === name);
            if (document) {
                document.states.labelingStatus = status;
            }
            if (state.currentDocument?.name === name) {
                state.currentDocument!.states.labelingStatus = status;
            }
        },
        clearCurrentDocument(state) {
            state.currentDocument = null;
        },
    },
    extraReducers: (builder) => {
        builder
            .addCase(resetPredictions, (state) => {
                state.documents.forEach((document) => {
                    document.states = { ...document.states, analyzingStatus: undefined };
                });
                if (state.currentDocument) {
                    state.currentDocument.states = { ...state.currentDocument.states, analyzingStatus: undefined };
                }
            })
            .addCase(addDocuments.pending, (state, action) => {
                const docs = action.meta.arg;
                docs.forEach((doc) => {
                    state.documents.push({
                        ...doc,
                        thumbnail: "",
                        numPages: 0,
                        currentPage: 0,
                        states: { loadingStatus: DocumentStatus.Loading },
                    });
                });
            })
            .addCase(setCurrentDocument.pending, (state, action) => {
                const document = action.meta.arg;
                const selectedDoc = state.documents.find((doc) => doc.name === document.name);
                if (selectedDoc) {
                    selectedDoc.states.loadingStatus = DocumentStatus.Loading;
                }
            })
            .addCase(setCurrentDocument.fulfilled, (state, action) => {
                const document: IDocument = {
                    ...action.payload.document,
                    states: {
                        ...action.payload.document.states,
                        loadingStatus: DocumentStatus.Loaded,
                    },
                };
                const selectedDoc = state.documents.find((doc) => doc.name === document.name);
                if (selectedDoc) {
                    selectedDoc.states.loadingStatus = DocumentStatus.Loaded;
                }
                state.currentDocument = document;
            })
            .addCase(setCurrentPage.fulfilled, (state, action) => {
                const { pageNumber } = action.payload;
                const document = state.documents.find((doc) => doc.name === state.currentDocument?.name);
                if (document) {
                    document.currentPage = pageNumber;
                }
                if (state.currentDocument) {
                    state.currentDocument.currentPage = pageNumber;
                }
            })
            .addCase(addDocuments.fulfilled, (state, action) => {
                const addedDocs = action.payload;
                const { documents } = state;
                addedDocs.forEach((doc) => {
                    const iDoc = documents.findIndex((d) => d.name === doc.name);
                    if (iDoc >= 0) {
                        documents[iDoc] = {
                            ...doc,
                            states: { ...documents[iDoc].states, loadingStatus: DocumentStatus.Loaded },
                        };
                    }
                });
            });
    },
});

export const { deleteDocument, setDocumentAnalyzingStatus, setDocumentLabelingStatus, clearCurrentDocument } =
    documentsSlice.actions;

export const reducer = documentsSlice.reducer;
