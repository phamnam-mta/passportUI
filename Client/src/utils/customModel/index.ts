import { SharedColors } from "@fluentui/react";
import { AnalyzeResultAdapterFactory } from "adapters/analyzeResultAdapter";
import { colors } from "consts/colors";
import { Field, FieldType, Label, Labels, LabelValue, LabelValueCandidate } from "models/customModels";
import { CustomModelAssetService } from "services/assetService/customModelAssetService";
import { FieldLocation, ICustomModelError } from "store/customModel/customModel";
import { IDocument } from "store/documents/documentsTypes";
import { FeatureCategory } from "view/components/imageMap/contracts";
import { createRegionIdFromPolygon } from "view/components/imageMap/utils";

export const buildRegionOrders = (analyzeResult) => {
    const analyzeResultAdapter = AnalyzeResultAdapterFactory.create(analyzeResult);
    const documentPage = analyzeResultAdapter.getDocumentPages();
    const regionOrders = {};
    documentPage.forEach((page) => {
        const { pageNumber, selectionMarks, width, height } = page;

        let order = 0;
        page.words.forEach((word) => {
            const { polygon } = word;
            if (polygon) {
                const normalizedPoints: number[] = [];
                for (let i = 0; i < polygon.length; i += 2) {
                    normalizedPoints.push(polygon[i] / width);
                    normalizedPoints.push(polygon[i + 1] / height);
                }
                const orderId = createRegionIdFromPolygon(normalizedPoints, pageNumber);
                regionOrders[orderId] = order++;
            }
        });

        if (selectionMarks) {
            selectionMarks.forEach((selectionMark) => {
                const { polygon } = selectionMark;
                if (polygon) {
                    const normalizedPoints: number[] = [];
                    for (let i = 0; i < polygon.length; i += 2) {
                        normalizedPoints.push(polygon[i] / width);
                        normalizedPoints.push(polygon[i + 1] / height);
                    }
                    const orderId = createRegionIdFromPolygon(normalizedPoints, pageNumber);
                    regionOrders[orderId] = order++;
                }
            });
        }
    });
    return regionOrders;
};

export const decodeLabelString = (label: string): string => {
    return label.replace(/~1/g, "/").replace(/~0/g, "~");
};

export const getFieldKeyFromLabel = (label: Label): string => {
    return decodeLabelString(label.label.split("/")[0]);
};

export const getTableFieldKeyFromLabel = (label: Label, fieldLocation: FieldLocation): string => {
    const labelIndex = fieldLocation === FieldLocation.field ? 1 : 2;
    return decodeLabelString(label.label.split("/")[labelIndex]);
};

export const getAllDocumentLabels = async (
    labels: Labels,
    documents: IDocument[],
    assetService: CustomModelAssetService
) => {
    const documentsWithoutLabels = documents.filter((document) => !labels[document.name]).map((doc) => doc.name);
    const addedLabels = await assetService.fetchAllDocumentLabels(documentsWithoutLabels);
    return { ...labels, ...addedLabels };
};

export const encodeLabelString = (label: string): string => {
    // Encoding for "/" since it will be used in table labeling result.
    return label.replace(/~/g, "~0").replace(/\//g, "~1");
};

export const getColorByFieldKey = (colorForFields: Record<string, string>[], fieldKey: string) => {
    const getDynamicKey = (obj: Record<string, string>) => Object.keys(obj)[0];

    const colorMap = colorForFields.find((color) => getDynamicKey(color) === fieldKey);
    return colorMap ? colorMap[fieldKey] : SharedColors.gray10;
};

export const getFieldColor = (fields: Field[], key: string): string => {
    const index = fields.findIndex((field) => field.fieldKey === key);
    const colorCounts = colors.length;

    if (index !== -1) {
        return colors[index % colorCounts];
    } else {
        return SharedColors.gray10;
    }
};

export const uniqueByKeepFirst = (array, key) => {
    let seen = new Set();
    return array.filter((item) => {
        let k = key(item);
        return seen.has(k) ? false : seen.add(k);
    });
};

export const makeLabelValue = (labelValueCandidate: LabelValueCandidate) => {
    const { page, text, boundingBoxes } = labelValueCandidate;
    return { page, text, boundingBoxes };
};

export const getOrder = (currentOrders: { [key: string]: number }, orderId: string): number => {
    const orderNumber = currentOrders[orderId];
    if (orderNumber !== undefined) {
        return orderNumber;
    }
    return 0;
};

export const compareOrder = (a: LabelValue, b: LabelValue, currentOrders, currentPage) => {
    const order1 = getOrder(currentOrders, createRegionIdFromPolygon(a.boundingBoxes[0], currentPage));
    const order2 = getOrder(currentOrders, createRegionIdFromPolygon(b.boundingBoxes[0], currentPage));
    return order1 >= order2 ? 1 : -1;
};

export const makeError = (name: string, message: string): ICustomModelError => {
    return { name, message };
};

export const validateAssignment = (candidates: LabelValueCandidate[], field: Field) => {
    switch (field.fieldType) {
        case FieldType.Signature:
            if (candidates.length !== 1 || candidates[0].category !== FeatureCategory.DrawnRegion) {
                throw makeError("Label assignment warning", "Signature field only supports one draw region per field.");
            }
            break;
        case FieldType.SelectionMark:
            if (
                candidates.length !== 1 ||
                !(
                    candidates[0].category === FeatureCategory.Checkbox ||
                    candidates[0].category === FeatureCategory.DrawnRegion
                )
            ) {
                throw makeError(
                    "Label assignment warning",
                    "SelectionMark field only supports one draw region or checkbox per field."
                );
            }
            break;
        default:
            // Text or table cell
            if (candidates.length > 1 && candidates.some((c) => c.category === FeatureCategory.DrawnRegion)) {
                throw makeError(
                    "Label assignment warning",
                    "General text field only supports one single draw region, or one or more text boxes."
                );
            }
            break;
    }
};

export const getUnusedFieldColor = (colorForFields: Record<string, string>[]) => {
    const getDynamicValue = (obj: Record<string, string>) => Object.values(obj)[0];

    const colorCounts = colors.length;
    const usedColors = colorForFields.map((field) => getDynamicValue(field));
    const unusedColors = colors.filter((color) => !usedColors.includes(color));

    if (colorForFields.length >= colorCounts && unusedColors.length === 0) {
        const initialUsageCount = colors.reduce((acc, color) => ({ ...acc, [color]: 0 }), {});
        const usageCount = usedColors.reduce((result, value) => ({ ...result, [value]: (result[value] || 0) + 1 }), {});

        const totalUsageCount = { ...initialUsageCount, ...usageCount };
        const colorCycleRound = Math.ceil(colorForFields.length / colorCounts);

        const leastUsedColorIndex = (Object.values(totalUsageCount) as number[]).findIndex(
            (count: number) => count < colorCycleRound
        );

        return colors[leastUsedColorIndex];
    }

    return unusedColors[0];
};
