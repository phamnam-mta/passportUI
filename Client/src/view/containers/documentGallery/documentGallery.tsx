import * as React from "react";
import { connect, ConnectedProps } from "react-redux";
import { Text } from "@fluentui/react";
import { DocumentPreviewList } from "./components/documentPreviewList";
import { setCurrentDocument, deleteDocument, clearCurrentDocument } from "store/documents/documents";
import { IDocument, DocumentStatus } from "store/documents/documentsTypes";
import { ApplicationState } from "store";
import MessageModal from "view/components/messageModal/messageModal";

import "./documentGallery.scss";

interface IDocumentGalleryProps {
    hideAddButton?: boolean;
    onDocumentDeleted?: (document: IDocument) => void;
    shouldConfirmDeleteDocument?: boolean;
}

type DocumentGalleryProps = ConnectedProps<typeof connector> & IDocumentGalleryProps;

interface IDocumentGalleryState {
    displayedDocuments: IDocument[];
    isDeleteModelOpen: boolean;
    documentToDeleted?: {
        name: string;
        index: number;
    };
}

export class DocumentGallery extends React.PureComponent<DocumentGalleryProps, IDocumentGalleryState> {
    constructor(props) {
        super(props);
        this.state = {
            displayedDocuments: props.documents,
            isDeleteModelOpen: false,
            documentToDeleted: undefined,
        };
    }

    public componentDidUpdate(prevProps) {
        const { documents, currentDocument, setCurrentDocument } = this.props;
        if (prevProps.documents !== documents) {
            this.setState({ displayedDocuments: documents });
        }
        if (
            documents &&
            documents[0] &&
            documents[0].states.loadingStatus === DocumentStatus.Loaded &&
            !currentDocument
        ) {
            setCurrentDocument(documents[0]);
        }
    }

    private handleDocumentClick = (name: string) => {
        const { documents, setCurrentDocument } = this.props;
        const selectedDoc = documents.find((doc) => doc.name === name);
        if (selectedDoc) {
            setCurrentDocument(selectedDoc);
        }
    };

    private setCurrentDocumentToNextOrPreviousDocument = (docIndexToDelete: number) => {
        const { documents, setCurrentDocument, clearCurrentDocument } = this.props;
        let nextDocument: IDocument;
        if (documents.length === 1) {
            clearCurrentDocument();
            return;
        }
        if (docIndexToDelete + 1 === documents.length) {
            nextDocument = documents[docIndexToDelete - 1];
        } else {
            nextDocument = documents[docIndexToDelete + 1];
        }
        setCurrentDocument(nextDocument);
    };

    private handleDocumentDelete = (docNameToDelete: string, index: number) => {
        const { documents, deleteDocument, currentDocument, onDocumentDeleted } = this.props;
        if (currentDocument?.name === docNameToDelete) {
            this.setCurrentDocumentToNextOrPreviousDocument(index);
        }
        const docToDelete = documents.find((doc) => doc.name === docNameToDelete);

        if (docToDelete) {
            deleteDocument(docToDelete);
            if (onDocumentDeleted) {
                onDocumentDeleted(docToDelete);
            }
        }
    };

    private handleConfirmDocumentDeletion = () => {
        const { name, index } = this.state.documentToDeleted!;
        this.handleDocumentDelete(name, index);
        this.setState({
            documentToDeleted: undefined,
            isDeleteModelOpen: false,
        });
    };

    private handleDocumentDeleteClicked = (name: string, index: number) => {
        this.setState({
            documentToDeleted: {
                name,
                index,
            },
            isDeleteModelOpen: true,
        });
    };

    public render() {
        const { currentDocument, shouldConfirmDeleteDocument } = this.props;
        const { displayedDocuments, isDeleteModelOpen, documentToDeleted } = this.state;
        return (
            <div className="document-gallery-container">
                <div className="document-gallery-list">
                    <DocumentPreviewList
                        currentDocument={currentDocument}
                        documents={displayedDocuments}
                        onDocumentClick={this.handleDocumentClick}
                        onDocumentDelete={
                            shouldConfirmDeleteDocument ? this.handleDocumentDeleteClicked : this.handleDocumentDelete
                        }
                    />
                </div>
                {isDeleteModelOpen && (
                    <MessageModal
                        isOpen={isDeleteModelOpen}
                        title="Delete Document"
                        body={
                            <Text variant="medium">
                                Labels and OCR results associated with this file will be deleted as well. Are you sure
                                to delete {documentToDeleted!.name},
                            </Text>
                        }
                        actionButtonText="Delete"
                        onActionButtonClick={this.handleConfirmDocumentDeletion}
                        onClose={() => this.setState({ isDeleteModelOpen: false })}
                    />
                )}
            </div>
        );
    }
}

const mapState = (state: ApplicationState) => ({
    documents: state.documents.documents,
    currentDocument: state.documents.currentDocument,
});
const mapDispatch = {
    setCurrentDocument,
    deleteDocument,
    clearCurrentDocument,
};

const connector = connect(mapState, mapDispatch);

export default connector(DocumentGallery);
