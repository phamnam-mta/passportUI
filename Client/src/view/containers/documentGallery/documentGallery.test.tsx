import * as React from "react";
import { DocumentStatus } from "store/documents/documentsTypes";
import { mockDocument, mockDocuments } from "utils/test";
import { DocumentGallery } from "./documentGallery";
import { shallow } from "enzyme";

jest.mock("utils/documentLoader");
jest.mock("utils", () => ({
    isSupportNewDocTypeEnabled: false,
}));

describe("<DocumentGallery />", () => {
    const myDocument = {
        name: "mock-image-1",
        type: "image/jpg",
        url: "mock-image-1.jpg",
        states: {
            loadingStatus: DocumentStatus.Loaded,
        },
    };

    const sampleDocument = {
        name: "mock-image-2",
        type: "image/jpg",
        url: "mock-image-2.jpg",
        states: {
            loadingStatus: DocumentStatus.Loaded,
        },
    };

    const mockUploadErrorMessage = "mock-upload-error-message";

    let baseProps;

    beforeEach(() => {
        baseProps = {
            documents: [myDocument, sampleDocument],
            currentDocument: null,
            setCurrentDocument: jest.fn(),
            deleteDocument: jest.fn(),
        };
    });

    describe("Rendering", () => {
        it("should match snapshot.", () => {
            const wrapper = shallow(<DocumentGallery {...baseProps} />);
            expect(wrapper).toMatchSnapshot();
        });

        it("should match snapshot, when documents are not provided.", () => {
            const props = { ...baseProps, documents: [] };
            const wrapper = shallow(<DocumentGallery {...props} />);
            expect(wrapper).toMatchSnapshot();
        });

        it("should match snapshot, when hideAddButton is set.", () => {
            const props = { ...baseProps, hideAddButton: true };
            const wrapper = shallow(<DocumentGallery {...props} />);
            expect(wrapper).toMatchSnapshot();
        });

        it("should match snapshot, when upload error is set.", () => {
            const wrapper = shallow(<DocumentGallery {...baseProps} />);
            wrapper.setState({
                uploadErrorMessage: mockUploadErrorMessage,
            });
            expect(wrapper).toMatchSnapshot();
        });

        it("should match snapshot, when delete error is set.", () => {
            const wrapper = shallow(<DocumentGallery {...baseProps} />);
            wrapper.setState({
                isDeleteModelOpen: true,
                documentToDeleted: { name: mockDocument.name, index: 0 },
            });

            expect(wrapper).toMatchSnapshot();
        });
    });

    describe("Event Handling", () => {
        it("should call setCurrentDocument, when document is clicked.", () => {
            const wrapper = shallow(<DocumentGallery {...baseProps} />);
            const documentClickedHandler = wrapper.find("DocumentPreviewList").prop("onDocumentClick") as any;

            documentClickedHandler(myDocument.name, 0);

            expect(baseProps.setCurrentDocument).toBeCalledTimes(1);
            expect(baseProps.setCurrentDocument).toBeCalledWith(myDocument);
        });

        it("should set the first sample document to the current document, when the document is updated.", () => {
            const wrapper = shallow(<DocumentGallery {...baseProps} />);
            wrapper.setProps({ documents: baseProps.documents });

            expect(baseProps.setCurrentDocument).toBeCalledTimes(1);
            expect(baseProps.setCurrentDocument).toBeCalledWith(myDocument);
        });

        it("should call deleteDocument and not triggered onDocumentClick to setCurrentDocument, when onDocumentDelete prop is triggered", () => {
            const wrapper = shallow(<DocumentGallery {...baseProps} />);
            const documentDeleteHandler = wrapper.find("DocumentPreviewList").prop("onDocumentDelete") as any;

            documentDeleteHandler(myDocument.name, 0);

            expect(baseProps.deleteDocument).toBeCalledTimes(1);
            expect(baseProps.deleteDocument).toBeCalledWith(myDocument);
            expect(baseProps.setCurrentDocument).not.toBeCalled();
        });

        it("should setCurrentDocument to next document, when onDocumentDelete prop is triggered and current document is to delete", () => {
            const mockDocToDelete = mockDocuments[0];
            const mockNextDocument = mockDocuments[1];
            const props = { ...baseProps, documents: mockDocuments, currentDocument: mockDocToDelete };
            const wrapper = shallow(<DocumentGallery {...props} />);
            const documentDeleteHandler = wrapper.find("DocumentPreviewList").prop("onDocumentDelete") as any;

            documentDeleteHandler(mockDocToDelete.name, 0);

            expect(baseProps.deleteDocument).toBeCalledTimes(1);
            expect(baseProps.deleteDocument).toBeCalledWith(mockDocToDelete);
            expect(baseProps.setCurrentDocument).toBeCalledTimes(1);
            expect(baseProps.setCurrentDocument).toBeCalledWith(mockNextDocument);
        });

        it("should setCurrentDocument to previous document, when current document is the document to delete and the last document in documentGallery", () => {
            const mockMyDocuments = new Array(3).fill(null).map((_, index) => ({
                name: `mock-image-${index}`,
                type: "image/jpg",
                url: `mock-image-${index}.jpg`,
                states: {
                    loadingStatus: DocumentStatus.Loaded,
                },
            }));
            const lastDocumentIndex = mockMyDocuments.length - 1;
            const mockDocToDelete = mockMyDocuments[lastDocumentIndex];
            const mockPreviousDocument = mockMyDocuments[lastDocumentIndex - 1];
            const props = { ...baseProps, documents: mockMyDocuments, currentDocument: mockDocToDelete };
            const wrapper = shallow(<DocumentGallery {...props} />);
            const documentDeleteHandler = wrapper.find("DocumentPreviewList").prop("onDocumentDelete") as any;

            documentDeleteHandler(mockDocToDelete.name, mockMyDocuments.length - 1);

            expect(baseProps.deleteDocument).toBeCalledTimes(1);
            expect(baseProps.deleteDocument).toBeCalledWith(mockDocToDelete);
            expect(baseProps.setCurrentDocument).toBeCalledTimes(1);
            expect(baseProps.setCurrentDocument).toBeCalledWith(mockPreviousDocument);
        });

        it("should call deleteDocument and not triggered onDocumentClick to setCurrentDocument, when onDocumentDelete prop is triggered", () => {
            const props = {
                ...baseProps,
                documents: [mockDocument],
                onDocumentDeleted: jest.fn(),
            };

            const wrapper = shallow(<DocumentGallery {...props} />);
            wrapper.setState({
                documentToDeleted: { name: mockDocument.name, index: 0 },
                isDeleteModelOpen: true,
            });
            const documentDeleteInStorageHandler = wrapper.find("MessageModal").prop("onActionButtonClick") as any;

            documentDeleteInStorageHandler();

            expect(props.onDocumentDeleted).toBeCalledTimes(1);
            expect(props.onDocumentDeleted).toBeCalledWith(mockDocument);
            expect(wrapper.state("documentToDeleted")).toBe(undefined);
            expect(wrapper.state("isDeleteModelOpen")).toBe(false);
        });
    });
});
