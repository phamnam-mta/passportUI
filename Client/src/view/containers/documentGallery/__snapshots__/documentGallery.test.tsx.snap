// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DocumentGallery /> Rendering should match snapshot, when delete error is set. 1`] = `
<div
  className="document-gallery-container"
>
  <div
    className="document-gallery-list"
  >
    <DocumentPreviewList
      currentDocument={null}
      documents={
        Array [
          Object {
            "name": "mock-image-1",
            "states": Object {
              "loadingStatus": "Loaded",
            },
            "type": "image/jpg",
            "url": "mock-image-1.jpg",
          },
          Object {
            "name": "mock-image-2",
            "states": Object {
              "loadingStatus": "Loaded",
            },
            "type": "image/jpg",
            "url": "mock-image-2.jpg",
          },
        ]
      }
      onDocumentClick={[Function]}
      onDocumentDelete={[Function]}
    />
  </div>
  <MessageModal
    actionButtonText="Delete"
    body={
      <Text
        variant="medium"
      >
        Labels and OCR results associated with this file will be deleted as well. Are you sure to delete 
        test.pdf
        ,
      </Text>
    }
    isOpen={true}
    onActionButtonClick={[Function]}
    onClose={[Function]}
    title="Delete Document"
  />
</div>
`;

exports[`<DocumentGallery /> Rendering should match snapshot, when documents are not provided. 1`] = `
<div
  className="document-gallery-container"
>
  <div
    className="document-gallery-list"
  >
    <DocumentPreviewList
      currentDocument={null}
      documents={Array []}
      onDocumentClick={[Function]}
      onDocumentDelete={[Function]}
    />
  </div>
</div>
`;

exports[`<DocumentGallery /> Rendering should match snapshot, when hideAddButton is set. 1`] = `
<div
  className="document-gallery-container"
>
  <div
    className="document-gallery-list"
  >
    <DocumentPreviewList
      currentDocument={null}
      documents={
        Array [
          Object {
            "name": "mock-image-1",
            "states": Object {
              "loadingStatus": "Loaded",
            },
            "type": "image/jpg",
            "url": "mock-image-1.jpg",
          },
          Object {
            "name": "mock-image-2",
            "states": Object {
              "loadingStatus": "Loaded",
            },
            "type": "image/jpg",
            "url": "mock-image-2.jpg",
          },
        ]
      }
      onDocumentClick={[Function]}
      onDocumentDelete={[Function]}
    />
  </div>
</div>
`;

exports[`<DocumentGallery /> Rendering should match snapshot, when upload error is set. 1`] = `
<div
  className="document-gallery-container"
>
  <div
    className="document-gallery-list"
  >
    <DocumentPreviewList
      currentDocument={null}
      documents={
        Array [
          Object {
            "name": "mock-image-1",
            "states": Object {
              "loadingStatus": "Loaded",
            },
            "type": "image/jpg",
            "url": "mock-image-1.jpg",
          },
          Object {
            "name": "mock-image-2",
            "states": Object {
              "loadingStatus": "Loaded",
            },
            "type": "image/jpg",
            "url": "mock-image-2.jpg",
          },
        ]
      }
      onDocumentClick={[Function]}
      onDocumentDelete={[Function]}
    />
  </div>
</div>
`;

exports[`<DocumentGallery /> Rendering should match snapshot. 1`] = `
<div
  className="document-gallery-container"
>
  <div
    className="document-gallery-list"
  >
    <DocumentPreviewList
      currentDocument={null}
      documents={
        Array [
          Object {
            "name": "mock-image-1",
            "states": Object {
              "loadingStatus": "Loaded",
            },
            "type": "image/jpg",
            "url": "mock-image-1.jpg",
          },
          Object {
            "name": "mock-image-2",
            "states": Object {
              "loadingStatus": "Loaded",
            },
            "type": "image/jpg",
            "url": "mock-image-2.jpg",
          },
        ]
      }
      onDocumentClick={[Function]}
      onDocumentDelete={[Function]}
    />
  </div>
</div>
`;
